src/sbvr-api/permissions.ts

```diff
@@ -430,17 +476,116 @@ const generateConstrainedAbstractSql = (
 	const collapsedPermissionFilters = collapsePermissionFilters(
 		permissionFilters,
 	);
-	_.set(odata, ['tree', 'options', '$filter'], collapsedPermissionFilters);
 
-	const translated = uriParser.translateUri({
-		method: 'GET',
-		resourceName,
+	return collapsedPermissionFilters;
+};
+
+const constrainedPermissionError = new PermissionError();
+const generateConstrainedAbstractSql = (
+	permissionsLookup: PermissionLookup,
+	actionList: PermissionCheck,
+	vocabulary: string,
+	resourceName: string,
+) => {
+	const abstractSQLModel = sbvrUtils.getAbstractSqlModel({
 		vocabulary,
-		odataBinds: odata.binds,
-		odataQuery: odata.tree,
-		values: {},
 	});
-	const abstractSqlQuery = _.clone(translated.abstractSqlQuery!);
+	const odata = memoizedParseOdata(`/${resourceName}`);
+
+	const collapsedPermissionFilters = buildODataPermission(
+		permissionsLookup,
+		actionList,
+		vocabulary,
+		resourceName,
+		odata,
+	);
+
+	_.set(odata, ['tree', 'options', '$filter'], collapsedPermissionFilters);
+
+	const lambdaAlias = randomstring.generate(20);
+	let inc = 0;
+
+	// We need to trace the processed resources, to be able to break
+	// permissions circles.
+	const canAccessTrace: string[] = [resourceName];
+	const odata2AbstractSQL = new OData2AbstractSQL(abstractSQLModel, {
+		canAccess: (compiler: OData2AbstractSQL, property: AnyObject): any => {
+			// remove method property so that we won't loop back here again at this point
+			delete property.method;
+
+			if (!compiler.defaultResource) {
+				throw new Error(`No resource selected in AST.`);
+			}
+
+			const targetResource = compiler.NavigateResources(
+				compiler.defaultResource,
+				property.name,
+			);
+
+			const targetResourceName = sqlNameToODataName(
+				targetResource.resource.name,
+			);
+
+			if (_.findIndex(canAccessTrace, v => targetResourceName === v) >= 0) {
+				// return a boolen subquery
+				return [
+					'Not',
+					[
+						'Exists',
+						['SelectQuery', ['Select', [['Alias', ['Number', 1], 'id']]]],
+					],
+				];
+			}
+
+			const parentOdata = memoizedParseOdata(`/${targetResourceName}`);
+
+			const collapsedParentPermissionFilters = buildODataPermission(
+				permissionsLookup,
+				actionList,
+				vocabulary,
+				targetResourceName,
+				parentOdata,
+			);
+
+			if (collapsedParentPermissionFilters === false) {
+				// We reuse a constant permission error here as it will be cached, and
+				// using a single error instance can drastically reduce the memory used
+				throw constrainedPermissionError;
+			}
+
+			const lId = `${lambdaAlias}+${inc}`;
+			inc = inc + 1;
+			rewriteSubPermissionBindings(
+				collapsedParentPermissionFilters,
+				compiler.bindVarsLength + compiler.extraBindVars.length,
+			);
+			convertToLamda(collapsedParentPermissionFilters, lId);
+			_.set(property, 'lambda', {
```

Comment:
This can be simplified to
```suggestion
			property.lambda = {
```
right?